jetbrains-core/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeWhispererFileCrawler.kt

```diff
@@ -40,51 +39,84 @@ interface FileCrawler {
      * @param psiFile psi of the test file we are searching with, e.g. MainTest.java
      * @return its source file e.g. Main.java, main.py or most relevant file if any
      */
-    fun findFocalFileForTest(psiFile: PsiFile): VirtualFile?
+    fun listUtgCandidate(psiFile: PsiFile): VirtualFile?
 
     /**
      * List files opened in the editors and sorted by file distance @see [CodeWhispererFileCrawler.getFileDistance]
+     * @return opened files and satisfy the following conditions
+     * (1) not the input file
+     * (2) with the same file extension as the input file has
+     * (3) non-test file which will be determined by [FileCrawler.isTestFile]
+     * (4) writable file
      */
-    fun listRelevantFilesInEditors(psiFile: PsiFile): List<VirtualFile>
+    fun listCrossFileCandidate(psiFile: PsiFile): List<VirtualFile>
+
+    /**
+     * Determine if the file given is test file or not based on its path and file name
+     */
+    fun isTestFile(virtualFile: VirtualFile, project: Project): Boolean
 }
 
 class NoOpFileCrawler : FileCrawler {
     override suspend fun listFilesImported(psiFile: PsiFile): List<VirtualFile> = emptyList()
 
     override fun listFilesUnderProjectRoot(project: Project): List<VirtualFile> = emptyList()
-    override fun findFocalFileForTest(psiFile: PsiFile): VirtualFile? = null
+    override fun listUtgCandidate(psiFile: PsiFile): VirtualFile? = null
 
     override fun listFilesWithinSamePackage(psiFile: PsiFile): List<VirtualFile> = emptyList()
 
-    override fun listRelevantFilesInEditors(psiFile: PsiFile): List<VirtualFile> = emptyList()
+    override fun listCrossFileCandidate(psiFile: PsiFile): List<VirtualFile> = emptyList()
+
+    override fun isTestFile(virtualFile: VirtualFile, project: Project): Boolean = false
 }
 
 abstract class CodeWhispererFileCrawler : FileCrawler {
     abstract val fileExtension: String
-    abstract val testFilenamePattern: Regex
     abstract val dialects: Set<String>
+    abstract val testFileNamingPatterns: List<Regex>
+
+    override fun isTestFile(virtualFile: VirtualFile, project: Project): Boolean {
+        val filePath = virtualFile.path
+
+        // if file path itself explicitly explains the file is under test sources
+        if (TestSourcesFilter.isTestSources(virtualFile, project) ||
+            filePath.contains("""test/""") ||
+            filePath.contains("""tst/""") ||
+            filePath.contains("""tests/""")
+        ) {
+            return true
+        }
+
+        // no explicit clue from the file path, use regexes based on naming conventions
+        return testFileNamingPatterns.any { it.matches(virtualFile.name) }
+    }
```

Comment:
moved from `CodeWhispererFileContextProvider` and further strenghen it by checking file name